vmdb/app/models/ems_refresh/parsers/foreman.rb

```diff
@@ -0,0 +1,116 @@
+module EmsRefresh
+  module Parsers
+    class Foreman
+      # we referenced a record that does not exist in the database
+      attr_accessor :missing_key
+
+      def self.provisioning_inv_to_hashes(inv)
+        new.provisioning_inv_to_hashes(inv)
+      end
+
+      def self.configuration_inv_to_hashes(inv)
+        new.configuration_inv_to_hashes(inv)
+      end
+
+      def provisioning_inv_to_hashes(inv)
+        result = {}
+        ids = {}
+        result[:customization_scripts] = customization_scripts_inv_to_hash(inv[:media], inv[:ptables])
+        add_ids(ids, result[:customization_scripts])
+        result[:operating_system_flavors] = operating_system_flavors_inv_to_hashes(inv[:operating_systems], ids)
+        result
+      end
+
+      def configuration_inv_to_hashes(inv)
+        result = {}
+        ids = {}
+        result[:configuration_profiles] = configuration_profile_inv_to_hashes(inv[:hostgroups],
+                                                                              inv[:operating_system_flavors],
+                                                                              inv[:customization_scripts])
+        add_ids(ids, result[:configuration_profiles])
+        result[:configured_systems] = configured_system_inv_to_hashes(inv[:hosts], ids, inv[:operating_system_flavors])
+        result[:missing_key] = true if missing_key
+        result
+      end
+
+      def media_inv_to_hashes(media)
+        media.collect do |m|
+          {
+            "manager_ref" => "medium:#{m["id"]}",
+            "type"        => "CustomizationScriptMedium",
+            "name"        => m["name"]
+          }
+        end
+      end
+
+      def ptables_inv_to_hashes(ptables)
+        ptables.collect do |m|
+          {
+            "manager_ref" => "ptable:#{m["id"]}",
+            "type"        => "CustomizationScriptPtable",
+            "name"        => m["name"]
+          }
+        end
+      end
+
+      def customization_scripts_inv_to_hash(media, ptables)
+        media_inv_to_hashes(media) + ptables_inv_to_hashes(ptables)
+      end
+
+      def operating_system_flavors_inv_to_hashes(flavors_inv, ids)
+        flavors_inv.collect do |os|
+          {
+            "manager_ref"           => "operating_system:#{os["id"]}",
```

Comment:
This too should just be the id...We should only qualify the ids where we have to, like in the customization scripts table where we blend objects.  Even so, I know I asked this before, but are those ids in the customization scripts table likely to be unique even across types, or could they overlap.